src/org/pentaho/reporting/platform/plugin/output/CachingPageableHTMLOutput.java

```diff
@@ -419,4 +433,16 @@ private CacheKeyException( final Throwable cause ) {
       super( cause );
     }
   }
+
+  private int getReportTotalRows( final String key ) {
+    final Map<String, Serializable> metaData = getCachedMetaData( key );
+    if ( metaData != null ) {
+      final Object reportTotalRows = metaData.get( REPORT_ROWS );
+      if ( reportTotalRows != null ) {
+        return (int) reportTotalRows;
+      }
+    }
+    return 0;
```

Comment:
If there is no valid result, I feel it might be not so nice to pretend you have one. Zero rows is a valid result if your table-model returned has no rows (SELECT * FROM x WHERE FALSE). So returning "-1" here might help keep the API's response clean.

If you return "0" here, you actually set the row-limit in the UI to zero (see RowLimitControl, line 110). If you return -1, you have a way of detecting this kind of bad state and you can leave the UI unchanged; and hopefully with the next full generation you get a valid result.

(You would only get to this point if the cache go cleared out between your first call to get the report-content and the second call here where you get the metadata. Would not happen in our implementation, but as the interfaces are public for OEMs, they can create their own backends which then may evict content at random points in time. The only way to avoid that race condition would be to combine the reading of the content and the metadata into one single operation so that either both items are available or none is. But lets not start with such a rewrite, lets handle the error in a predictable way instead.)